libs/sdk-ui-dashboard/src/model/state/filterContext/filterContextReducers.ts

```diff
@@ -1,14 +1,111 @@
 // (C) 2021 GoodData Corporation
 
 import { Action, CaseReducer, PayloadAction } from "@reduxjs/toolkit";
+import { v4 as uuidv4 } from "uuid";
+import {
+    DateFilterGranularity,
+    DateFilterType,
+    DateString,
+    FilterContextItem,
+    IDashboardAttributeFilter,
+    IDashboardDateFilter,
+    isDashboardAttributeFilter,
+    isDashboardDateFilter,
+} from "@gooddata/sdk-backend-spi";
+import invariant from "ts-invariant";
 import { FilterContextState } from "./filterContextState";
+import { IAttributeElements } from "@gooddata/sdk-model";
 
 type FilterContextReducer<A extends Action> = CaseReducer<FilterContextState, A>;
 
+const generateFilterLocalIdentifier = (): string => uuidv4().replace(/-/g, "");
+
 const setFilterContext: FilterContextReducer<PayloadAction<any>> = (state, action) => {
-    state.filterContext = action.payload;
+    state.filterContext = {
+        ...action.payload,
+        // make sure attribute filters always have localId
+        filters: action.payload.filters?.map((filter: FilterContextItem) =>
+            isDashboardAttributeFilter(filter)
+                ? {
+                      attributeFilter: {
+                          ...filter.attributeFilter,
+                          localIdentifier:
+                              filter.attributeFilter.localIdentifier ?? generateFilterLocalIdentifier(),
+                      },
+                  }
+                : filter,
+        ),
+    };
+};
+
+const upsertDateFilter: FilterContextReducer<
+    PayloadAction<
+        | { readonly type: "allTime" }
+        | {
+              readonly type: DateFilterType;
+              readonly granularity: DateFilterGranularity;
+              readonly from?: DateString | number;
+              readonly to?: DateString | number;
+          }
```

Comment:
can you extract this to a named type pls?

Reply:
Extracted all the payload types to named types.